File: example/java/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/example/model/Book.java
package org.babyfish.jimmer.sql.example.model;

import org.babyfish.jimmer.sql.*;
import org.babyfish.jimmer.sql.example.model.common.TenantAware;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.util.List;

@Entity
public interface Book extends TenantAware {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @Key
    int edition();

    BigDecimal price();

    @Nullable // Null property, Java API requires this annotation, but kotlin API does not
    @ManyToOne
    BookStore store();

    @ManyToMany(orderedProps = {
            @OrderedProp("firstName"),
            @OrderedProp("lastName")
    })
    @JoinTable(
            name = "BOOK_AUTHOR_MAPPING",
            joinColumnName = "BOOK_ID",
            inverseJoinColumnName = "AUTHOR_ID"
    )
    List<Author> authors();

    @OneToMany(mappedBy = "book", orderedProps = @OrderedProp("index"))
    List<Chapter> chapters();

    // -----------------------------
    // Optional properties
    // -----------------------------

    // Optional property `storeId`
    // If this property is deleted, please add `BookInput.Mapper.toBookStore(Long)`
    @IdView
    Long storeId();

    // Optional property `authorIds`
    // If this property is deleted, please add `BookInputMapper.toAuthor(Long)`
    @IdView("authors")
    List<Long> authorIds();
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/classic/book/Book.kt
package org.babyfish.jimmer.sql.kt.model.classic.book

import org.babyfish.jimmer.sql.*
import org.babyfish.jimmer.sql.kt.model.classic.store.BookStore
import org.babyfish.jimmer.sql.kt.model.classic.author.Author
import java.math.BigDecimal
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Positive
import javax.validation.constraints.PositiveOrZero

typealias ID = Long

/**
 * The Book Entity, 100% immutable
 */
@Entity
interface Book {

    /**
     * The id property
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: ID

    /**
     * The name property, 100% immutable
     */
    @Key
    @get:NotEmpty(message = "The book name cannot be empty")
    val name: String

    /**
     * The edition property, 100% immutable
     */
    @Key
    val edition: @PositiveOrZero Int

    /**
     * The price property, 100% immutable
     */
    val price: @Positive BigDecimal

    /**
     * The store property, 100% immutable
     */
    @ManyToOne
    val store: BookStore?

    /**
     * The authors property, 100% immutable
     */
    @ManyToMany
    @JoinTable(deletedWhenEndpointIsLogicallyDeleted = true)
    val authors: List<Author>

    /**
     * The id view property of `store`, 100% immutable
     */
    @IdView
    val storeId: Long?

    /**
     * The id view property of `authorIds`, 100% immutable
     */
    @IdView("authors")
    val authorIds: List<Long>
}